docs/environment_variables.md
# Environment Variables

The following environment variables can be used to configure different application:

## nac_api

### Shared
Environment variables shared between external and internal mode.

| Variable | Type | Default | Description |
| :--- | :--- | :--- | :--- |
| `POSTGRES_USER` | `string` | "cnaas" | Postgres username |
| `POSTGRES_PASSWORD` | `string` | "cnaas" | Postgres password |
| `POSTGRES_SERVER` | `string` | "nac_postgres" | Postgres server |
| `POSTGRES_PORT` | `int` | 5432 | Postgres port |
| `POSTGRES_DB` | `string` | "nac" | Postgres db name |
| `LOGGING` | `string | int` | `INFO` | Set to `DEBUG` to enable verbose debug logging |
| `ENVIRONMENT` | `string` | `local` when run locally, `production` for the docker image. | Sets the running environment. Mainly used in security functions |

### External specific
| Variable | Type | Default | Description |
| :--- | :--- | :--- | :--- |
| `OIDC_CLIENT_ID` | `string` | "cnaas-nac" | OIDC client id |
| `OIDC_CLIENT_SECRET` | `string` | "" | OIDC client secret |
| `OIDC_DISCOVERY_URL` | `string` | "" | OIDC discovery url |
| `OIDC_USERINFO_ATTRIBUTE` | `string` | "userinfo" | OIDC userinfo attribute |
| `OIDC_USERNAME_ATTRIBUTE` | `string` | "preferred_username" | userinfo username attribute |
| `OIDC_GROUPS_ATTRIBUTE` | `string` | "roles" | userinfo groups attribute |
| `OIDC_ADMIN_GROUP` | `string` | "admins" | Which OIDC group should map to the admin group |
| `OIDC_ADMIN_USERS` | `list[string]` | [] | Used to manually set users to the internal admin group |
| `SECRET_KEY` | `string` | "replace_this_with_a_secure_random_string" | Used for session middleware |
| `JWT_EXPIRATION_MINUTES` | `int` | 60 | How long the jwt access token should be valid |
| `FRONTEND_CALLBACK_URL` | `string` | "/#/auth-callback" | Frontend callback during OIDC login |

### Internal specific
| Variable | Type | Default | Description |
| :--- | :--- | :--- | :--- |
| `PRUNING_DISABLED` | `bool` | False | Disable pruning by settings this to true |
| `ENDPOINT_MAB_DISCOVERED_RETENTION_DAYS` | `int` | 30 | How long to keep discovered MAB endpoints before they are pruned |
| `ENDPOINT_MAB_PENDING_RETENTION_DAYS` | `int` | 30 | How long to keep pending MAB endpoints before they are pruned |
| `ENDPOINT_MAB_REJECTED_RETENTION_DAYS` | `int` | 30 | How long to keep rejected MAB endpoints before they are pruned |
| `ENDPOINT_EAP_REJECTED_RETENTION_DAYS` | `int` | 30 | How long to keep rejected EAP endpoints before they are pruned |
| `ENDPOINT_MAB_AUTHORIZED_RETENTION_DAYS` | `int` | 90 | How long to keep authorized MAB endpoints before they are pruned |
| `ENDPOINT_EAP_AUTHORIZED_RETENTION_DAYS` | `int` | 90 | How long to keep authorized EAP endpoints before they are pruned |
| `RADACCT_RETENTION_DAYS` | `int` | 90 | How long to keep radacct(sessions) before they are pruned |
| `RADPOSTAUTH_RETENTION_DAYS` | `int` | 90 | How long to keep radpostauth(authentications) before they are pruned |

## nac_radius

| Variable | Type | Default | Description |
| :--- | :--- | :--- | :--- |
| `PRE_START_BASE_FOLDER` | `string` | "/etc/raddb" | Where the init-script should look for template files during bootup. |
| `RADIUS_PROXY_CONFIGS` | `list (object)` | `[]` | See: [RADIUS_PROXY_CONFIGS](#radius_proxy_configs). |
| `RADIUS_EAP_CONFIGS` | `list (object)` | `[]` | See: [RADIUS_EAP_CONFIGS](#radius_eap_configs). |
| `LDAP_CONFIGS` | `list (object)` | `[]` | See: [LDAP_CONFIGS](#ldap_configs). |


### Examples

#### RADIUS_PROXY_CONFIGS
```bash
RADIUS_PROXY_CONFIGS: |
    [
        {
            "realm": "example.org",
            "pool_name": "" # Optional
            "pool_type": "" # Optional, default fail-over
            "servers": [
                {
                    "name": "radius",
                    "ipaddr": "radius.example.org",
                    "port": "1812", # Optional, default 1812
                    "secret": "testing123"
                }
            ]
        }
    ]
```

#### RADIUS_EAP_CONFIGS
!!! warning "WIP"
    Not yet fully implemented.

```bash
RADIUS_EAP_CONFIGS: |
    [
        {
            "name": "",
            "domain": "",
            "priv_key": "",
            "cert": "",
            "ca": "",
            "check_crl": false,
            "crl_url": ""
        }
    ]
```

#### LDAP_CONFIGS
```bash
LDAP_CONFIGS: |
    [
        {
            "ad_username_attr": "mail", # Optional, default sAMAccountName
            "ad_member_attr": "", # Optional, default: memberOf
            "ad_member_filter":""  # Optionally used instead of ad_member_attr
            "ad_domain": "example.org", 
            "ad_server": "server1.example.org", # Optional hostname other than the domain name
            "ad_username": "", # Bind user
            "ad_password": "",
            "ad_base_dn": ""
        }
    ]
```

test/eos-a1_startup_config.txt
!
no aaa root
!
username admin privilege 15 role network-admin secret sha512 $6$cpGmCQItlNlsZSTv$/Pl3AN1CoUO1yiy3P7.eEP9ZzqO5r5fJbWBaeL7ZUOeEHIAjxwbj4xooYPKEtxJSlwLEbYEgNJuAYKhcvsNE4.
!
management api http-commands
   no shutdown
!
no service interface inactive port-id allocation disabled
!
transceiver qsfp default-mode 4x10G
!
service routing protocols model multi-agent
!
hostname eos-a1
!
spanning-tree mode mstp
!
system l1
   unsupported speed action error
   unsupported error-correction action error
!
vlan 13
   name quarantine
!
vlan 14
   name vlan_14
!
management api gnmi
   transport grpc default
!
management api netconf
   transport ssh default
!
radius-server attribute 32 include-in-access-req hostname
!
radius-server host 10.100.2.1 key 7 120D0004060202037B7977
!
aaa group server radius cnaas-nac
   server 10.100.2.1
!
aaa authentication dot1x default group cnaas-nac
aaa accounting dot1x default start-stop group cnaas-nac
!
interface Vlan14
   ip address 192.168.0.1/24
!
interface Ethernet1
   description DOT1X
   spanning-tree portfast
   spanning-tree bpduguard enable
   dot1x pae authenticator
   dot1x authentication failure action traffic allow vlan 13
   dot1x reauthentication
   dot1x port-control auto
   dot1x mac based authentication
   dot1x timeout reauth-period 60
   dot1x eapol authentication failure fallback mba timeout 5
!
interface Ethernet2
   description DOT1X
   spanning-tree portfast
   spanning-tree bpduguard enable
   dot1x pae authenticator
   dot1x authentication failure action traffic allow vlan 13
   dot1x reauthentication
   dot1x port-control auto
   dot1x mac based authentication
   dot1x timeout reauth-period 60
   dot1x eapol authentication failure fallback mba timeout 5
!
!interface Ethernet3
!   description DOT1X
!   spanning-tree portfast
!   spanning-tree bpduguard enable
!   dot1x pae authenticator
!   dot1x authentication failure action traffic allow vlan 13
!   dot1x reauthentication
!   dot1x port-control auto
!   dot1x mac based authentication
!   dot1x timeout reauth-period 60
!   dot1x eapol authentication failure fallback mba timeout 5
!
!interface Ethernet4
!   description DOT1X
!   spanning-tree portfast
!   spanning-tree bpduguard enable
!   dot1x pae authenticator
!   dot1x authentication failure action traffic allow vlan 13
!   dot1x reauthentication
!   dot1x port-control auto
!   dot1x mac based authentication
!   dot1x timeout reauth-period 60
!   dot1x eapol authentication failure fallback mba timeout 5
!
!interface Ethernet5
!   description DOT1X
!   spanning-tree portfast
!   spanning-tree bpduguard enable
!   dot1x pae authenticator
!   dot1x authentication failure action traffic allow vlan 13
!   dot1x reauthentication
!   dot1x port-control auto
!   dot1x mac based authentication
!   dot1x timeout reauth-period 60
!   dot1x eapol authentication failure fallback mba timeout 5
!
interface Management0
   ip address 10.100.2.11/24
!
no ip routing
!
ip route 0.0.0.0/0 10.100.2.1
!
dot1x system-auth-control
dot1x dynamic-authorization
!
end

src/cnaas_nac/core/settings.py
# from enum import Enum
from enum import Enum
from typing import Annotated, Any, Optional

from pydantic import Field, computed_field, field_validator
from pydantic_settings import BaseSettings

import logging


class EnvironmentOption(Enum):
    LOCAL = "local"
    STAGING = "staging"
    PRODUCTION = "production"


class Settings(BaseSettings):
    POSTGRES_USER: str = "cnaas"
    POSTGRES_PASSWORD: str = "cnaas"
    POSTGRES_SERVER: str = "nac_postgres"
    POSTGRES_PORT: int = 5432
    POSTGRES_DB: str = "nac"

    POSTGRES_SYNC_PREFIX: str = "postgresql://"
    POSTGRES_ASYNC_PREFIX: str = "postgresql+asyncpg://"

    POSTGRES_URL: Optional[str] = None

    @computed_field  # type: ignore[prop-decorator]
    @property
    def POSTGRES_URI(self) -> str:
        return f"{self.POSTGRES_USER}:{self.POSTGRES_PASSWORD}@{self.POSTGRES_SERVER}:{self.POSTGRES_PORT}/{self.POSTGRES_DB}"

    OIDC_CLIENT_ID: str = "cnaas-nac"
    OIDC_CLIENT_SECRET: str = ""
    OIDC_DISCOVERY_URL: str = ""
    OIDC_USERINFO_ATTRIBUTE: str = "userinfo"
    OIDC_USERNAME_ATTRIBUTE: str = "preferred_username"
    OIDC_GROUPS_ATTRIBUTE: str = "roles"
    OIDC_ADMIN_GROUP: str = "admins"
    # Used to manually set users to the internal admin group
    OIDC_ADMIN_USERS: list[str] = []

    SECRET_KEY: str = "replace_this_with_a_secure_random_string"
    JWT_EXPIRATION_MINUTES: int = 60

    FRONTEND_CALLBACK_URL: str = "/#/auth-callback"

    ENVIRONMENT: EnvironmentOption = EnvironmentOption.LOCAL

    PRUNING_DISABLED: bool = False

    ENDPOINT_MAB_DISCOVERED_RETENTION_DAYS: Annotated[int, Field(gt=0)] = 30
    ENDPOINT_MAB_PENDING_RETENTION_DAYS: Annotated[int, Field(gt=0)] = 30

    ENDPOINT_MAB_REJECTED_RETENTION_DAYS: Annotated[int, Field(gt=0)] = 30
    ENDPOINT_EAP_REJECTED_RETENTION_DAYS: Annotated[int, Field(gt=0)] = 30

    ENDPOINT_MAB_AUTHORIZED_RETENTION_DAYS: Annotated[int, Field(gt=0)] = 90
    ENDPOINT_EAP_AUTHORIZED_RETENTION_DAYS: Annotated[int, Field(gt=0)] = 90

    RADACCT_RETENTION_DAYS: Annotated[int, Field(gt=0)] = 90
    RADPOSTAUTH_RETENTION_DAYS: Annotated[int, Field(gt=0)] = 90

    LOGGING: str | int = logging.INFO

    @field_validator("LOGGING", mode="before")
    @classmethod
    def setup_logging(cls, v: Any) -> int:
        if not v:
            return logging.INFO
        if v not in logging._nameToLevel:
            return logging.INFO
        return logging._nameToLevel.get(v, logging.INFO)


settings = Settings()
